File: clvm/core_ops.py
from typing import Tuple, TypeVar

from typing_extensions import Never

from .EvalError import EvalError
from .SExp import SExp

from .costs import (
    IF_COST,
    CONS_COST,
    FIRST_COST,
    REST_COST,
    LISTP_COST,
    EQ_BASE_COST,
    EQ_COST_PER_BYTE,
)


_T_SExp = TypeVar("_T_SExp", bound=SExp)


def op_if(args: _T_SExp) -> Tuple[int, _T_SExp]:
    if args.list_len() != 3:
        raise EvalError("i takes exactly 3 arguments", args)
    r = args.rest()
    if args.first().nullp():
        return IF_COST, r.rest().first()
    return IF_COST, r.first()


def op_cons(args: _T_SExp) -> Tuple[int, _T_SExp]:
    if args.list_len() != 2:
        raise EvalError("c takes exactly 2 arguments", args)
    return CONS_COST, args.first().cons(args.rest().first())


def op_first(args: _T_SExp) -> Tuple[int, _T_SExp]:
    if args.list_len() != 1:
        raise EvalError("f takes exactly 1 argument", args)
    return FIRST_COST, args.first().first()


def op_rest(args: _T_SExp) -> Tuple[int, _T_SExp]:
    if args.list_len() != 1:
        raise EvalError("r takes exactly 1 argument", args)
    return REST_COST, args.first().rest()


def op_listp(args: SExp) -> Tuple[int, SExp]:
    if args.list_len() != 1:
        raise EvalError("l takes exactly 1 argument", args)
    return LISTP_COST, args.true if args.first().listp() else args.false


def op_raise(args: SExp) -> Never:
    if args.list_len() == 1 and not args.first().listp():
        raise EvalError("clvm raise", args.first())
    else:
        raise EvalError("clvm raise", args)


def op_eq(args: SExp) -> Tuple[int, SExp]:
    if args.list_len() != 2:
        raise EvalError("= takes exactly 2 arguments", args)
    a0 = args.first()
    a1 = args.rest().first()
    if a0.pair or a1.pair:
        raise EvalError("= on list", a0 if a0.pair else a1)
    b0 = a0.as_atom()
    assert b0 is not None
    b1 = a1.as_atom()
    assert b1 is not None
    cost = EQ_BASE_COST
    cost += (len(b0) + len(b1)) * EQ_COST_PER_BYTE
    return cost, (args.true if b0 == b1 else args.false)

File: src/core_ops.rs
use super::node::Node;
use super::types::{EvalErr, Reduction};

impl Node {
    pub fn first(&self) -> Result<Node, EvalErr> {
        match self.as_pair() {
            Some((a, _b)) => Ok(a),
            _ => self.node_err("first of non-cons"),
        }
    }

    pub fn rest(&self) -> Result<Node, EvalErr> {
        match self.as_pair() {
            Some((_a, b)) => Ok(b),
            _ => self.node_err("rest of non-cons"),
        }
    }
}

pub fn op_if(args: &Node) -> Result<Reduction, EvalErr> {
    let cond = args.first()?;
    let mut chosen_node = args.rest()?;
    if cond.nullp() {
        chosen_node = chosen_node.rest()?;
    }
    Ok(chosen_node.first()?.into())
}

pub fn op_cons(args: &Node) -> Result<Reduction, EvalErr> {
    let a1 = args.first()?;
    let a2 = args.rest()?.first()?;
    Ok(Node::pair(&a1, &a2).into())
}

pub fn op_first(args: &Node) -> Result<Reduction, EvalErr> {
    Ok(args.first()?.first()?.into())
}

pub fn op_rest(args: &Node) -> Result<Reduction, EvalErr> {
    Ok(args.first()?.rest()?.into())
}

pub fn op_listp(args: &Node) -> Result<Reduction, EvalErr> {
    match args.first()?.as_pair() {
        Some((_first, _rest)) => Ok(Node::from(1).into()),
        _ => Ok(Node::null().into()),
    }
}

pub fn op_raise(args: &Node) -> Result<Reduction, EvalErr> {
    args.err("clvm raise")
}

pub fn op_eq(args: &Node) -> Result<Reduction, EvalErr> {
    let a0 = args.first()?;
    let a1 = args.rest()?.first()?;
    if let Some(s0) = a0.as_atom() {
        if let Some(s1) = a1.as_atom() {
            if s0 == s1 {
                return Ok(Node::blob_u8(&[1]).into());
            }
            return Ok(Node::null().into());
        }
    }
    args.err("= on list")
}
